### ISSUE TITLE
sokol-shdc: text/binary output for stand-alone version

### ISSUE THREAD

**Issue by code-disaster** [floooh/sokol-tools#14]
Some thoughts I came up with while tinkering with my "sokol_shdc + cr.h live reload" hack.

In one of my private projects, I've created a glslang + spirv-cross cli tool, doing pretty much the same as the stand-alone version of sokol-shdc, only without all the _really fancy_ stuff. :)

Instead of going NIH once more, I'd like to just adopt your solution. But, as you may have noticed, live shader reloading is very up on my priority list. As fun & cool as the topic of live reloading of C code seems to be, it may not be the most ... "pragmatic" approach.

So, the idea is:

- embed shader code generation in the C/C++ project, either doing the full deal or just for the type definitions
- allow stand-alone version to dump result of shader compilation as text or binary file(s)
- (for development work,) treat shader code more "traditionally", as an asset, so it can be updated, compiled and reloaded while the game is running

Possible usage:

If a new argument, e.g. `--raw`, is specified, use `--output` as a base path. Write compilation results with some naming convention for vertex/fragment shader, and a file extensions based on `--slang`. Or, use separate arguments to explicitly specify vertex/fragment output paths.

I can give it a try myself if you like (and if we can agree on the cli).

**Comment by floooh** [comment]
Yes... I've been thinking about "multiple output formats" as well, but with Oryol in mind.

In a way, sokol.cc is only the first output-backend (but without going all too "architecture-crazy", the backend would just be selected with a simple if-else at the end of the main function, and each backend gets a command-line argument to activate it.

For instance:
- --sokol (which doesn't exist yet) would activate the sokol.cc output backend
- --oryol for a new backend for Oryol implemented in an oryol.cc source
- --raw ... for what you suggest.

I wasn't happy with the somewhat non-descriptive "--raw" at first, but I couldn't come up with anything better ;)

I think the "magic" you describe to derive the other details (like file extensions) from existing arguments is the right way (maybe only add new args when the "magic way" gets too confusing).

If you want to start with an implementation for --raw, go ahead :) (in the future I think it's better anyway  to add more specialized output backends instead of trying to "bend" existing output code for new requirements).

**Comment by floooh** [comment]
...one small correction maybe... since we may soon run out of meaningful "single-letter-command-line-options", it's probably better to have a single command line argument to select the output-backend...

Maybe something like: ```--backend sokol```, ```--backend raw```...

However, ```-b``` is already taken...maybe you come up with a better name ;)

**Comment by code-disaster** [comment]
`-f --format=[sokol,bare,oryol]`

**Comment by floooh** [comment]
perfect!

**Comment by code-disaster** [comment]
There's a first, _barely_ (ha!) working version pushed at https://github.com/code-disaster/sokol-tools/tree/bare-output-format. Probably could need some more error checking love.

**Comment by mattiasljungstrom** [comment]
Thanks for implementing this! 

I'm working on a setup in my code where I can reload binary shaders, but still compile them with the "sokol" format during the build. 

@floooh : Would it be possible to add the filename where the shader came from into the sokol output? (without path?). I could work around the issue by naming my shaders carefully or writing this into the code, but automatic stuff is always nicer. :)

**Comment by floooh** [comment]
Without the path shouldn't be a problem. If absolute path would be required it might be a problem if people generate the file "in source" and commit into version control.

Do you want to provide a PR @mattiasljungstrom ? Even if it's incomplete it would be useful to continue the discussion in a PR's thread with concrete source code to talk about :)

**Comment by mattiasljungstrom** [comment]
Would it be ok to add a new const char field to sg_shader_desc after .label? Or should I add a new static const with the filename, like this:
static const char* SHADERNAME_filename = "shader";

The first approach would be nicer for me, but I can understand if you don't want to add things to sg_shader_desc. Or do you have any other suggestions?

**Comment by floooh** [comment]
I would prefer the ```static const char*``` version.

Similar to this, but without the shader-language name at the end:

https://github.com/floooh/sokol-tools/blob/57a12281f78e5426e6a900e12b43a29ad8af5029/src/shdc/sokol.cc#L435

So basically:

```cpp
L("static const char {}{}_shader_source = \"{}\";\n", mod_prefix(inp), prog.name, filename);
```

**Comment by mattiasljungstrom** [comment]
Created a pull request here: https://github.com/floooh/sokol-tools/pull/18

**Comment by mattiasljungstrom** [comment]
Note: If there are multiple shaders in the same .glsl file, it looks like the bare format will write them all to the same filename. I'll look into it.

**Comment by code-disaster** [comment]
You are right, I didn't consider both mod_prefix() and prog.name for naming the output files.

**Comment by mattiasljungstrom** [comment]
Don't merge my previous pull request, I'm going to suggest another change that will make it unnecessary for my use.

**Comment by mattiasljungstrom** [comment]
Please review this pull request instead (https://github.com/floooh/sokol-tools/pull/19). It adds support for multiple shaders in the same .glsl file and will output a fs/vs pair for each program.

Since it writes files that are named after the shader program name I wouldn't need my previous additions.

**Comment by code-disaster** [comment]
PR is looking good. I also like that you can now just give `--output <path/to/folder/>`, and if this folder exists, it cleanly drops all files there w/o complaining.

**Comment by floooh** [comment]
...oh wait... forget my comment in the PR :D

**Comment by floooh** [comment]
...so this can be closed without merging: https://github.com/floooh/sokol-tools/pull/18

...and this is can be merged? https://github.com/floooh/sokol-tools/pull/19

I've been looking at the 2 PRs first, not at this comment thread, so I'm a bit confused ATM ;)

**Comment by mattiasljungstrom** [comment]
yes exactly, sorry about that, I should have closed #18.

**Comment by floooh** [comment]
closing since #19 has been merged
